packages/yarnpkg-pnp/sources/esm-loader/loader.ts

```diff
@@ -0,0 +1,192 @@
+import {NativePath, PortablePath}          from '@yarnpkg/fslib';
+import {parse, init}                       from 'cjs-module-lexer';
+import fs                                  from 'fs';
+import moduleExports                       from 'module';
+import path                                from 'path';
+import {fileURLToPath, pathToFileURL, URL} from 'url';
+
+import * as nodeUtils                      from '../loader/nodeUtils';
+import {PnpApi}                            from '../types';
+
+function isValidURL(str: string) {
+  try {
+    new URL(str);
+    return true;
+  } catch {
+    return false;
+  }
+}
+
+const builtins = new Set([...moduleExports.builtinModules]);
+
+const pathRegExp = /^(?![a-zA-Z]:[\\/]|\\\\|\.{0,2}(?:\/|$))((?:node:)?(?:@[^/]+\/)?[^/]+)\/*(.*|)$/;
+
+async function exists(path: string) {
+  try {
+    await fs.promises.access(path, fs.constants.R_OK);
+    return true;
+  } catch { }
+  return false;
+}
+
+export async function resolve(
+  originalSpecifier: string,
+  context: any,
+  defaultResolver: any
+) {
+  const {findPnpApi} = (moduleExports as unknown as { findPnpApi?: (path: NativePath) => null | PnpApi });
+  if (!findPnpApi)
+    return defaultResolver(originalSpecifier, context, defaultResolver);
+
+  let specifier = originalSpecifier;
+  let validURL: boolean | undefined;
+  if (builtins.has(specifier) || (validURL = isValidURL(specifier))) {
+    if (!validURL || pathToFileURL(specifier).protocol !== `file:`) {
+      return defaultResolver(originalSpecifier, context, defaultResolver);
+    } else {
+      specifier = fileURLToPath(specifier);
+    }
+  }
+
+  const {parentURL, conditions = []} = context;
+
+  const issuer = parentURL ? fileURLToPath(parentURL) : process.cwd();
+
+  const pnpapi = findPnpApi(issuer) ?? (validURL ? findPnpApi(specifier) : null);
+  if (!pnpapi)
+    return defaultResolver(originalSpecifier, context, defaultResolver);
+
+  const dependencyNameMatch = specifier.match(pathRegExp);
+
+  let allowLegacyResolve = false;
+
+  if (dependencyNameMatch) {
+    const [, dependencyName, subPath] = dependencyNameMatch as [unknown, string, PortablePath];
+
+    // https://github.com/nodejs/node/blob/0996eb71edbd47d9f9ec6153331255993fd6f0d1/lib/internal/modules/esm/resolve.js#L686-L691
```

Comment:
General remark: the code could benefit from more comments explaining the reason for some parts of the implementation (edge cases, etc). For example here, I assume the comment would be something like this:

```suggestion
    // If the package.json doesn't list an `exports` field, Node will tolerate omitting the extension
    // https://github.com/nodejs/node/blob/0996eb71edbd47d9f9ec6153331255993fd6f0d1/lib/internal/modules/esm/resolve.js#L686-L691
```